### PR TITLE
Do less work computing diff-merge results in the common case.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/73424

### CODE REVIEW

File: src/Features/CSharpTest/InlineTemporary/InlineTemporaryTests.cs [hunk 1]
```diff
@@ -5769,7 +5769,7 @@ public class Goo
                     {
                         public void Bar()
                         {
-                            var target = new List&lt;object;gt>();
+                            var target = new List&lt;object&gt;();
```
Comment (CyrusNajmabadi) [inline src/Features/CSharpTest/InlineTemporary/InlineTemporaryTests.cs:5772]: tests were wonky.

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs [hunk 2]
```diff
@@ -19,88 +17,112 @@ namespace Microsoft.CodeAnalysis;
...
+        var unmergedChanges = new List<UnmergedDocumentChanges>();
+        for (int i = 1, n = newDocumentsAndHashes.Count; i < n; i++)
         {
+            var siblingNewDocument = newDocumentsAndHashes[i].newDocument;
+            var siblingOldDocument = oldSolution.GetRequiredDocument(siblingNewDocument.Id);
+
             appliedChanges = await AddDocumentMergeChangesAsync(
-                oldSolution.GetDocument(documentId),
-                newSolution.GetDocument(documentId),
-                [.. appliedChanges],
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs:84]: this was converting an IA into a list, that was never mutated.  made the input/output entirely IA bases.

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs [hunk 3]
```diff
@@ -19,88 +17,112 @@ namespace Microsoft.CodeAnalysis;
...
+            {
+                newDocumentsAndHashes = [];
+                filePathToNewDocumentsAndHashes.Add(filePath, newDocumentsAndHashes);
+            }
+
+            var newText = await newDocument.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
+            var newContentHash = newText.GetContentHash();
+            // Ignore any linked documents that we have the same contents as.  
+            if (newDocumentsAndHashes.Any(t => t.newContentHash.SequenceEqual(newContentHash)))
+                continue;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs:41]: this is the meat of the change.  we just ignore siblings with the same hash as one we've already seen.  that means it has the same text, so no point trying to fetch that text (and subtexts) to figure out the diff/merge info.

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs [hunk 4]
```diff
@@ -19,88 +17,112 @@ namespace Microsoft.CodeAnalysis;
...
+            Contract.ThrowIfTrue(newDocumentsAndHashes.Count == 0);
 
-            // Ensure the first document in the group is the first in the list of 
-            var allLinkedDocuments = documentInNewSolution.GetLinkedDocumentIds().Add(documentInNewSolution.Id);
-            if (allLinkedDocuments.Length == 1)
-            {
+            // Don't need to do anything if this document has no linked siblings.
+            var firstNewDocument = newDocumentsAndHashes[0].newDocument;
+            var relatedDocuments = newSolution.GetRelatedDocumentIds(firstNewDocument.Id);
+            if (relatedDocuments.Length == 1)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs:56]: this is the next meat part.  if we've only truly come up with one change to a doc, then we don't even have to do any diff/merge work.  we can just take that change and apply to all linked files.

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs [hunk 5]
```diff
@@ -19,88 +17,112 @@ namespace Microsoft.CodeAnalysis;
...
+            if (newDocumentsAndHashes.Count == 1)
             {
-                var mergeGroupResult = await MergeLinkedDocumentGroupAsync(allLinkedDocuments, linkedDocumentsWithChanges, sessionInfo, mergeConflictHandler, cancellationToken).ConfigureAwait(false);
-                linkedFileMergeResults.Add(mergeGroupResult);
-                mergedText = mergeGroupResult.MergedSourceText;
+                // The file has linked siblings, but we collapsed down to only one actual document change.  Ensure that
+                // any linked files have that same content as well.
+                var firstSourceText = await firstNewDocument.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
+                updatedSolution = updatedSolution.WithDocumentTexts(
+                    relatedDocuments.SelectAsArray(d => (d, firstSourceText)));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs:65]: this is the next meat part. if we've only truly come up with one change to a doc, then we don't even have to do any diff/merge work. we can just take that change and apply to all linked files.

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs:56]: note: could potentially lift this up to the dictionary building loop.  i will consider that.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs:56]: i don't see the GetContentHash being expensive at all (it's 0.2% of hte total trace).  so preventing this by lifting this to the dictionary loop doesn't really help.

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs [hunk 6]
```diff
@@ -19,87 +17,112 @@ namespace Microsoft.CodeAnalysis;
...
+            if (!filePathToNewDocumentsAndHashes.TryGetValue(filePath, out var newDocumentsAndHashes))
+            {
+                newDocumentsAndHashes = [];
+                filePathToNewDocumentsAndHashes.Add(filePath, newDocumentsAndHashes);
+            }
+
+            var newText = await newDocument.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
+            var newContentHash = newText.GetContentHash();
+            // Ignore any linked documents that we have the same contents as.  
+            if (newDocumentsAndHashes.Any(t => t.newContentHash.SequenceEqual(newContentHash)))
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs:40]: >if (newDocumentsAndHashes.Any(t => t.newContentHash.SequenceEqual(newContentHash))) [](http://example.com/codeflow?start=12&length=83)

Probably not significant, but isn't there a closure and IEnumerable allocation on every loop iteration here? We've got a custom Any impl for ImmutableArray which avoids these, is there one somewhere for List?

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs [hunk 7]
```diff
@@ -19,87 +17,112 @@ namespace Microsoft.CodeAnalysis;
...
-            .GroupBy(d => oldSolution.GetDocument(d).FilePath, StringComparer.OrdinalIgnoreCase);
+        using var _ = PooledDictionary<string, List<(Document newDocument, ImmutableArray<byte> newContentHash)>>.GetInstance(out var filePathToNewDocumentsAndHashes);
+        foreach (var documentId in solutionChanges.GetProjectChanges().SelectMany(p => p.GetChangedDocuments()))
+        {
+            var newDocument = newSolution.GetRequiredDocument(documentId);
+            var filePath = newDocument.FilePath;
+            Contract.ThrowIfNull(filePath);
 
-        var linkedFileMergeResults = new List<LinkedFileMergeResult>();
+            if (!filePathToNewDocumentsAndHashes.TryGetValue(filePath, out var newDocumentsAndHashes))
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs:31]: >TryGetValue [](http://example.com/codeflow?start=49&length=11)

nit: there isn't a GetOrAdd extension available to use that takes in a Func to get the value, is there?

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs:40]: yeah, it's not showing as significant her.e

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs]: done.

**Comment by CyrusNajmabadi** [comment]
@ToddGrun anything else?

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**